ebu_tt_live/bindings/validation.py

```diff
@@ -159,43 +194,223 @@ def _post_timing_set_attribute(self, attr_use):
+        self._pre_calculate_begin()
+
+    def _post_pop_begin(self):
+        if self._begin_timedelta is not None:
+            # We pushed on the stack it is time to pop it. It could probably be removed
+            # and replaced with self._begin_timedelta
+            begin_timedelta = self._semantic_dataset['timing_begin_stack'].pop()
+            self._semantic_dataset['timing_syncbase'] -= begin_timedelta
+
+    def _post_pop_end(self):
+        end_timedelta = None
+
+        if self._end_timedelta is not None:
+            # We pushed on the stack it is time to pop it
+            end_timedelta = self._semantic_dataset['timing_end_stack'].pop()
+
+        return end_timedelta
+
+    def _semantic_postprocess_timing(self, dataset, element_content):
+
+        self._post_pop_begin()
+        # This end timedelta is an absolute calculated value on the timeline. Not relative.
+        end_timedelta = self._post_pop_end()
+
+        # If the forward running part of the traversal could not assign an end time we can use the backwards route
+        # which is in a way similar to dynamic programming because we take the children computed times and take the
+        # maximum value from them. SPECIAL case: a single leaf element with an undefined end time renders the entire
+        # branch undefined
+
+        if end_timedelta is not None and self.computed_end_time is None \
+                or end_timedelta is None and self.computed_end_time is not None:
+            # This is just a simple sanity check. It should never be triggered.
+            # Should the calculation be changed this filters out an obvious source of error.
+            raise LogicError()
+
+        if self.computed_end_time is None:
+            # This requires calculation based on the timings in its children.
+
+            # All timing containers are complexTypes so we can call orderedContent safely
+            children = filter(lambda item: isinstance(item, TimingValidationMixin), [x.value for x in self.orderedContent()])
+            # Order of statements is important
+            if not children:
+                # This means we are in a timing container leaf.
+                if not self._semantic_dataset['timing_end_stack']:
+                    # Here we go an endless document. Pointless but for clarity's sake assign it explicitly to None.
+                    self._computed_end_time = None
+                else:
+                    # Try to assign it the last specified ancestor
+                    self._computed_end_time = self._semantic_dataset['timing_end_stack'][-1]
+            else:
+                children_computed_end_times = [item.computed_end_time for item in children]
+
+                if None in children_computed_end_times:
+                    # The endless document case propagates up
+                    self._computed_end_time = None
+                else:
+                    # Propagate the longest end time among the children
```

Comment:
Might be worth adding a comment that this is where the `endsync="all"` semantic is coded.